### ISSUE TITLE
Possible to Mock Method Calls?

### ISSUE THREAD

**Issue by Midacts** [pester/Pester#965]
Is there a way to `mock` methods?

(Get-Date).AddDays(1)

To be able to mock the ",AddDays" call.

**Comment by dlwyatt** [comment]
No, but you can wrap that method call in a PowerShell function and then mock that function.  It's annoying, but it's what we've got to work with.

```posh
function Add-Days([datetime]$time, [int] $numDays) {
    return $time.AddDays($numDays)
}

mock Add-Days { Do-Whatever }
```

**Comment by Midacts** [comment]
I worked with someone in the powershell.slack.com #testing channel to mock `Get-WmiObject` calls (chrisdent maybe?).
The below code worked.

Is that something that can be reproduced for other method calls?
Or was this just something that would only work for specific instances.

If it would work for others, i am not sure how to figure out what what in the `Add-Member` delete value.

``` PowerShell
# Actual code
$CompObject = Get-WmiObject -Query "SELECT * FROM ..." -ComputerName <SERVER> -NameSpace "ROOT\..."
$CompObject.PSBase.Delete()

# Pester mock
Mock Get-WmiObject { 
	$Object = New-Object System.Management.ManagementObject
	$Object.PSBase | Add-Member Delete -MemberType ScriptMethod -Force -Value { $Script:deleteCalled = $true }
	$Object
}
```

**Comment by nohwnd** [comment]
@Midacts I think it was me as well. I think this is a path worth looking into, but I would need more examples to see if it is actually that useful.

**Comment by johlju** [comment]
We use that method extensively in the [unit tests over at SqlServerDsc resource module](https://github.com/PowerShell/SqlServerDsc/tree/dev/Tests/Unit). Example for unit tests below for a resource that is currently in PR. It also counts number of calls that is made to the method.

https://github.com/SQLHorizons/xSQLServer/blob/f26967cb1ab48200cf6c9ab9ea8c6e3bf53ccbea/Tests/Unit/MSFT_SqlServerDatabaseMail.Tests.ps1#L111-L132

**Comment by Midacts** [comment]
Sadly, the example I have posted is the best/only working example I have.
I have run into scenarios where I have used method calls in my functions, which is why I was prompted to ask the question.

I was curious if this is actually a valid option (and if we have some "formula" to make this repeatable as I can't figure out how to accomplish this on my own). Or if the only option to mock things like this is to follow Dave's suggestion.
